sv.c

```diff
@@ -5795,6 +5795,30 @@ Perl_newSV(pTHX_ const STRLEN len)
     }
     return sv;
 }
+
+/*
+=for apidoc newSVpvz
+
+Creates a new SV sized with enough preallocated space to hold a string of
+C<len> bytes (plus a trailing NUL).
+
+The SV is initialized with a zero-length string with C<SvPOK> set.
+
+The reference count for the new SV is set to 1.
+
+=cut
+*/
+SV *
+Perl_newSVpvz(pTHX_ const STRLEN len)
+{
+    SV *sv = newSV_type(SVt_PV);
+    sv_grow_fresh(sv, len + 1);
+    SvPOK_only(sv);
+    SvPVX(sv)[0] = '\0';
```

Comment:
Instead of it merely being preallocated, what if it actually was a string of that length? Either zeroed out or undefined value? Because when you preallocate it like this you probably don't want to have to mess around with `SvCUR` right after.

Reply:
As it stands, the new SV will have `SvCUR == 0`, so it's up to the caller to do the right thing. If it returned `SvCUR == len` then it would be good to zero out the entire buffer so it's determined; I wouldn't like the idea of having uninitialised bytes visible there.

Reply:
It could use `sv_setpv_freshbuf(sv)` to initialize (which handles tainting possibilities).  But why does that function call `SvPOK_only_UTF8`?  I think it is irrelevant for a fresh PV, but it seems to me that it would be more robust to clear the flag.

I'm inclined to say something like

> Creates a new SV initialized with a zero-length string, and sized with enough
preallocated space to hold a string of len bytes (plus a trailing NUL)
without needing to grow.

Reply:
That does raise a good point, this function could use a flags argument for `SVf_UTF8`

Reply:
> It could use sv_setpv_freshbuf(sv) to initialize (which handles tainting possibilities).

It's not clear to me why you would want to use that.

>  But why does that function call SvPOK_only_UTF8? I think it is irrelevant for a fresh PV, but it seems to me that it would be more robust to clear the flag.

It disables typeflags except POK and UTF8, that does sound like the right thing to do in that particular case.

Reply:
I don't understand your point.  This is a fresh SV.  Creating it does not turn on the UTF-8 flag.  Therefore there is no need to preserve it being on.  Keeping it at zero is a no-op.  And hence using a construct that preserves it makes me stop and question what is going on here?  And the answer after wasting my time is nothing is going on.  Or do I misunderstand?

And I don't see the need for a flags argument.  If we knew what is going to be placed into the PV, we would be using something like 'newSVpv()` and its kin.  I would be using this function when I don't know what the string is going to be, much less if it is going to be UTF-8 or not, but I do have a guess as to how big to make it so as to avoid in most circumstances the need for additional reallocs.  Changing a flag setting is a lot cheaper than a malloc potentially is.

Maybe my situation is so specialized that a special function isn't warranted, but I have wanted this very functionality in the past.

Reply:
> Maybe my situation is so specialized that a special function isn't warranted, but I have wanted this very functionality in the past.

I think we just have different situations in mind.

Reply:
Could you elaborate on the situation you are thinking of?